ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java

```diff
@@ -64,17 +70,29 @@ public DataStreamClientImpl(RaftPeer raftServer,
     this.orderedStreamAsync = new OrderedStreamAsync(dataStreamClientRpc, properties);
   }
 
-  class DataStreamOutputImpl implements DataStreamOutput {
+  public class DataStreamOutputImpl implements DataStreamOutput {
     private long streamId = 0;
     private long messageId = 0;
+    private RaftClientRequest request;
```

Comment:
Let's create the header and send it out in the constructor.
```suggestion
    private final RaftClientRequest header;
    private final CompletableFuture<DataStreamReply> headerFuture;
```